src/nvme/nbft.c

```diff
@@ -0,0 +1,947 @@
+#define SECFLAG_AUTHENTICATION_POLICY_LIST_REQ			 0x0010
+#define SECFLAG_AUTHENTICATION_POLICY_LIST_RSVD			 0x0018
+#define SECFLAG_SECURE_CHANNEL_NEGOTIATION			0x0060
+#define SECFLAG_SECURE_CHANNEL_NEGOTIATION_NOT_SUP		 0x0000
+#define SECFLAG_SECURE_CHANNEL_NEGOTIATION_SUP			 0x0020
+#define SECFLAG_SECURE_CHANNEL_NEGOTIATION_REQ			 0x0040
+#define SECFLAG_SECURE_CHANNEL_NEGOTIATION_RSVD			 0x0060
+#define SECFLAG_SECURITY_POLICY_LIST				0x0180
+#define SECFLAG_SECURITY_POLICY_LIST_NOT_PRES			 0x0000
+#define SECFLAG_SECURITY_POLICY_LIST_PRES			 0x0080
+#define SECFLAG_SECURITY_POLICY_LIST_PRES_ADMINSET		 0x0100
+#define SECFLAG_SECURITY_POLICY_LIST_RSVD			 0x0180
+#define SECFLAG_CIPHER_SUITES_RESTRICTED_BY_POLICY		0x0200
+#define SECFLAG_AUTH_DH_GROUPS_RESTRICTED_BY_POLICY_LIST	0x0400
+#define SECFLAG_SECURE_HASH_FUNCTIONS_POLICY_LIST		0x0800
+
+enum secret_type {
+	SECRET_TYPE_RESERVED,
+	SECRET_TYPE_REDFISH_HOST_INTERFACE_URI,
+};
+
+/*
+ * DISCOVERY DESCRIPTOR (Figure 24)
+ */
+struct __attribute__((__packed__)) raw_nbft_discovery {
+	__u8 structure_id;
+	__u8 flags;
+	__u8 index;
+	__u8 hfi_index;
+	__u8 security_index;
+	__u8 reserved1;
+	heap_obj discovery_ctrl_addr;
+	heap_obj discovery_controller_nqn;
+	__u8 reserved2[14];
+};
+
+#define DISCOVERYFLAG_VALID	0x01
+
+/*
+ *  End of NBFT ACPI table definitions
+ */
+#define MIN(a,b) (((a)<(b))?(a):(b))
+
+static __u8 csum(void *buffer, int length)
+{
+	int n;
+	__u8 sum = 0;
+
+	for (n = 0; n < length; n++) {
+		sum = (__u8)(sum + ((__u8 *)buffer)[n]);
+	}
+	return sum;
+}
+
+static void format_ip_addr(char *buf, size_t buflen, __u8 *addr) {
+	struct in6_addr *addr_ipv6;
+
+	addr_ipv6 = (struct in6_addr *)addr;
+	if (   addr_ipv6->s6_addr32[0] == 0
+	    && addr_ipv6->s6_addr32[1] == 0
+	    && (ntohl(addr_ipv6->s6_addr32[2]) == 0xffff) )
+		/* ipv4 */
+		inet_ntop(AF_INET, &(addr_ipv6->s6_addr32[3]), buf, buflen);
+	else
+		/* ipv6 */
+		inet_ntop(AF_INET6, addr_ipv6, buf, buflen);
+}
+
+static int in_heap(struct raw_nbft_header *header, heap_obj obj)
+{
+	if (obj.length == 0)
+		return 1;
+	if (obj.offset < header->heap_offset)
+		goto bad;
+	if (obj.offset > header->heap_offset + header->heap_length)
+		goto bad;
+	if (obj.offset + obj.length > header->heap_offset + header->heap_length)
+		goto bad;
+	return 1;
+bad:
+	return 0;
+}
+
+/*
+ *  Return transport_type string (NBFT Table 2)
+ */
+static char *trtype_to_string(__u8 transport_type)
+{
+	switch (transport_type) {
+		case 3:
+			return "tcp";
+			break;
+		default:
+			return "invalid";
+			break;
+	}
+}
+
+#define verify(condition, message)							\
+	if (!(condition)) {								\
+		nvme_msg(NULL, LOG_DEBUG, "file %s: " message "\n", nbft->filename);	\
+		return -EINVAL;								\
+	}
+
+static int __get_heap_obj(struct raw_nbft_header *header, const char *filename,
+			  const char *descriptorname, const char *fieldname,
+			  heap_obj obj, bool is_string,
+			  char **output)
+{
+	if (obj.length == 0) {
+		*output = NULL;
```

Comment:
Usually, if an error is returned the callee is not supposed to change anything. The caller is supposed to handle the error correctly.